### PR TITLE
use typescript in driver

### PR DESCRIPTION
**Pull request by Bkucera**
- this sets us up to use typescript in the driver, and converts a few files
- use webpack-preprocessor in driver cypress tests
- correct some `chai` usage e.g.: `expect(...).to.be.a.array`
- [x] wait on #5382 upgrade eslint-plugin-dev

- Closes NA

### PR Tasks

- [NA] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [NA] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [NA] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/src/cy/commands/actions/type.js [hunk 1]
```diff
@@ -111,7 +111,7 @@ module.exports = function (Commands, Cypress, cy, state, config) {
       }
 
       const isBody = options.$el.is('body')
-      const isTextLike = $dom.isTextLike(options.$el)
+      const isTextLike = $dom.isTextLike(options.$el[0])
```
Comment (jennifer-shehane) [inline packages/driver/src/cy/commands/actions/type.js]: How is this not changing behavior?

File: packages/driver/src/dom/elements.ts [hunk 2]
```diff
@@ -389,10 +430,11 @@ const isElement = function (obj) {
...
 
-const isFocusable = ($el) => {
-  return _.some(focusable, (sel) => {
-    return $el.is(sel)
-  })
+/**
+ * The element can be activeElement, recieve focus events, and also recieve keyboard events
+ */
+const isFocusable = ($el: JQuery<Element>) => {
+  return _.some(focusable, (sel) => $el.is(sel)) || (isElement($el[0]) && getTagName($el[0]) === 'html' && isContentEditable($el[0]))
```
Comment (brian-mann) [inline packages/driver/src/dom/elements.ts]: extract the right side to a named function - this one liner is way too much

File: packages/driver/src/dom/elements.ts [hunk 3]
```diff
@@ -405,8 +447,13 @@ const isW3CFocusable = (el) => {
   return isFocusable(wrap(el)) && isW3CRendered(el)
 }
 
-const isType = function ($el, type) {
-  const el = [].concat($jquery.unwrap($el))[0]
+const isType = function (el: HTMLInputElement | HTMLInputElement[] | HTMLButtonElement | HTMLButtonElement[] | JQuery<HTMLInputElement> | JQuery<HTMLButtonElement>, type) {
```
Comment (brian-mann) [inline packages/driver/src/dom/elements.ts]: can we name all these as a new common type?

File: packages/driver/src/cy/ensures.coffee [hunk 4]
```diff
@@ -117,7 +117,7 @@ create = (state, expect) ->
 
     # readonly can only be applied to input/textarea
     # not on checkboxes, radios, etc..
-    if $dom.isTextLike(subject) and subject.prop("readonly")
+    if $dom.isTextLike(subject[0]) and subject.prop("readonly")
```
Comment (Bkucera) [inline packages/driver/src/cy/ensures.coffee]: use `.get(0)` to make it clear this is a jquery obj

File: packages/driver/test/cypress/integration/commands/connectors_spec.coffee [hunk 5]
```diff
@@ -480,8 +480,8 @@ describe "src/cy/commands/connectors", ->
                 memo + num
               , 0
             math: {
-              sum: =>
-                @obj.sum.apply(@obj, arguments)
+              sum: (args...) =>
+                @obj.sum.apply(@obj, args)
```
Comment (Bkucera) [inline packages/driver/test/cypress/integration/commands/connectors_spec.coffee]: arguments is not available in arrow function, so this change was needed

File: packages/driver/ts/import-coffee.d.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+declare module '*.coffee' {
```
Comment (Bkucera) [inline packages/driver/ts/import-coffee.d.ts]: add comment why this file is needed: to make typescript ok with .coffee resolves

File: packages/driver/test/cypress/plugins/index.js [hunk 7]
```diff
@@ -1,9 +1,16 @@
+require('@packages/ts/register')
```
Comment (Bkucera) [inline packages/driver/test/cypress/plugins/index.js]: add comment: this is to be able to read in `.ts` webpack config

File: packages/network/index.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+if (process.env.CYPRESS_ENV !== 'production') {
```
Comment (Bkucera) [inline packages/network/index.ts]: so can we remove `index.js` now?

File: packages/runner/ts/import-coffee.d.ts [hunk 9]
```diff
@@ -0,0 +1,4 @@
+declare module '*.coffee' {
```
Comment (Bkucera) [inline packages/runner/ts/import-coffee.d.ts]: add comment explaining why this is needed

File: packages/network/index.ts (hunk 8 above)
Comment (Bkucera) [reply packages/network/index.ts]: No, build breaks

**Pull request merged**